docs/best-practices.md

```diff
@@ -0,0 +1,416 @@
+
+  private def filterPublicKeys(user: UserWithIdentities): Step[Seq[UserIdentity]] =
+    Sync[IO].pure(user.identities.filter(i => i.`type` == IdentityType.CryptoVerification))
+      .seal
+      .ensure(_.nonEmpty, LoginByCryptoVerificationResult.InvalidCredentials)
+
+  def login(
+      prevAuthContext: Option[AuthContext],
+      loginRequest: LoginByCryptoVerificationRequest
+  ): M[LoginByCryptoVerificationResult] =
+    (for {
+      _                   <- validatePrevAuthContext(prevAuthContext, loginRequest)
+      userWithIdentities  <- findUserWithIdentities(loginRequest)
+      publicKeyIdentities <- filterPublicKeys(userWithIdentities)
+      identity            <- verify(publicKeyIdentities, loginRequest)
+      result              <- loginResolve(loginRequest, userWithIdentities, prevAuthContext, identity)
+    } yield LoginByCryptoVerificationResult.Ok(result)).run
+}
+```
+
+---
+
+### 3.2 For-Comprehensions for Readability
+
+**Best Practice:** Favor for-comprehensions to flatten nested monadic operations and handle errors gracefully.
+
+**Example:** The following snippet from a permission service demonstrates how conditional logic and complex checks are expressed clearly:
+
+```scala
+def checkAccess(authContext: AuthContext, organizationId: OrganizationId): Step[Unit] =
+  membershipService(organizationId, authContext.id)
+    .seal
+    .attempt {
+      case OrganizationUserMembershipResult.Admin  => Left(AccessControlResult.Allowed)
+      case OrganizationUserMembershipResult.Member => Right(())
+      case OrganizationUserMembershipResult.None   => Left(AccessControlResult.NotAllowed)
+    }
+    .map(_ => ())
+```
+
+---
+
+## 4. Side Effect Management
+
+### 4.1 Controlled Side Effects with tap and flatTap
+
+**Best Practice:** Use operators like `tap`, `flatTap`, and `semiflatTap` to execute side effects without disrupting the main pipeline.
+
+**Example:** In the email confirmation service, logging is performed via side effects without altering the result:
+
+```scala
+userT.semiflatMap { case (method, user) =>
+  upsertAuthMethod(confirmMethod(method))
+    .map(_ => ConfirmResponse.Confirmed(issueTokenFor(user)))
+}
+.merge
+```
+
+### 4.2 Branching Side Effects Based on Monad State
+
+**Best Practice:** Differentiate side effects based on the success or error path for accurate logging and debugging.
+
+**Example:** In our permission list service, caching occurs only when a fresh state is computed:
+
+```scala
+calculateAndCacheCurrentPermissionsState(organizationId, authUserId).biSemiflatTap(
+  {
+    case OrganizationUserPermissionListResult.Ok(permissions) =>
+      cacheCurrentPermissionsState(organizationId, authUserId, permissions, Instant.now())
+    case _ =>
+      IO.unit
+  },
+  permissions => cacheCurrentPermissionsState(organizationId, authUserId, permissions, Instant.now()).map(_.permissions)
+)
+```
+
+---
+
+## 5. Testing Strategies
+
+### 5.1 Test All Possible Results
```

Comment:
While it's true, it doesn't apply to (only) sealed monad. It's just a general programming advice.